internal/config/pipeline.go

```diff
@@ -118,148 +166,244 @@ func (c *Config) PipelineProtos() ([]*pb.Pipeline, error) {
-			Name: pl.Name,
-			Owner: &pb.Pipeline_Project{
-				Project: &pb.Ref_Project{
-					Project: c.hclConfig.Project,
-				},
-			},
+		pipes, err := c.buildPipelineProto(pl)
+		if err != nil {
+			return nil, err
 		}
 
-		// TODO(briancain): Use the default ODR image for image_url so that it isn't required
-		// We do this already maybe in the ODR task launcher.
-		steps := make(map[string]*pb.Pipeline_Step)
-		for i, step := range pl.StepRaw {
-			s := &pb.Pipeline_Step{
-				Name:      step.Name,
-				DependsOn: step.DependsOn,
-				Image:     step.ImageURL, //TODO(briancain): actually use this when executing steps
-			}
+		result = append(result, pipes...)
+	}
 
-			// If no dependency was explictily set, we rely on the previous step
-			if i != 0 && len(step.DependsOn) == 0 {
-				s.DependsOn = []string{pl.StepRaw[i-1].Name}
-			}
+	// We should validate cycles across pipelines here
 
-			// We currently only support one kind of Step plugin. But in the future
-			// maybe this would be a switch on step.Type? Or maybe we get our
-			// own Step Eval func that returns the step proto instead and does the
-			// switches there.
-			// NOTE(briancain): This is what you'd change to support future Step plugins
-			switch step.Use.Type {
-			case "build":
-				var buildBody struct {
-					DisablePush bool `hcl:"disable_push,optional"`
-				}
+	return result, nil
+}
 
-				if diag := gohcl.DecodeBody(step.Use.Body, finalizeContext(c.ctx), &buildBody); diag.HasErrors() {
-					return nil, diag
-				}
+// buildPipelineProto will recursively translate an hclPipeline into a protobuf
+// Pipeline message.
+func (c *Config) buildPipelineProto(pl *hclPipeline) ([]*pb.Pipeline, error) {
+	var result []*pb.Pipeline
+	pipe := &pb.Pipeline{
+		Name: pl.Name,
+		Owner: &pb.Pipeline_Project{
+			Project: &pb.Ref_Project{
+				Project: c.hclConfig.Project,
+			},
+		},
+	}
 
-				s.Kind = &pb.Pipeline_Step_Build_{
-					Build: &pb.Pipeline_Step_Build{
-						DisablePush: buildBody.DisablePush,
-					},
-				}
-			case "deploy":
-				var deployBody struct {
-					Release bool `hcl:"release,optional"`
-				}
+	steps := make(map[string]*pb.Pipeline_Step)
+	for i, step := range pl.StepRaw {
+		s := &pb.Pipeline_Step{
+			Name:      step.Name,
+			DependsOn: step.DependsOn,
+			Image:     step.ImageURL,
+		}
+
+		// If no dependency was explictily set, we rely on the previous step
+		if i != 0 && len(step.DependsOn) == 0 {
+			s.DependsOn = []string{pl.StepRaw[i-1].Name}
+		}
 
-				if diag := gohcl.DecodeBody(step.Use.Body, finalizeContext(c.ctx), &deployBody); diag.HasErrors() {
-					return nil, diag
+		// We have an embeded pipeline for this step. This can either be an hclPipeline
+		// defined directly in the step, or a pipeline reference to another pipeline
+		// defined else where. If this is a ref, the raw hcl for the pipeline should
+		// be empty.
+		if step.PipelineRaw != nil {
+			// TODO(briancain): It shouldn't be valid to have a "project/pipeline"
```

Comment:
Does this mean that a nested pipeline should not be able to refer to a pipeline in another project?

Reply:
Ah, so this comment is no longer valid! This was still around when I had external pipeline references as

```hcl
step "external-pipeline" {
  pipeline "project/pipeline" { }
  // This shouldn't be valid to have this too
  pipeline "test" {
    // config
  }
}
```

But got some feedback that the pipeline ref was too magical, so now you reference a pipeline with a step:

```hcl
step "external-pipeline-ref" {
  use "pipeline" {
    project = "test"
    name    = "my-pipeline"  
  }
}
```

So yeah, I'll remove this comment!